File: internal/service/weather_service.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"

	"mcp-weather-server/internal/models"
	"mcp-weather-server/pkg/logger"
)

// WeatherService interface for weather data operations
type WeatherService interface {
	GetWeatherData(ctx context.Context, city string) (*models.WeatherData, error)
}

// weatherService implements WeatherService interface
type weatherService struct {
	apiKey string
	client *http.Client
	logger logger.Logger
}

// NewWeatherService creates a new weather service instance
func NewWeatherService(apiKey string, logger logger.Logger) WeatherService {
	return &weatherService{
		apiKey: apiKey,
		client: &http.Client{
			Timeout: 10 * time.Second,
		},
		logger: logger,
	}
}

// GetWeatherData fetches weather data from OpenWeatherMap API
func (w *weatherService) GetWeatherData(ctx context.Context, city string) (*models.WeatherData, error) {
	// Create request URL
	baseURL := "https://api.openweathermap.org/data/2.5/weather"
	params := url.Values{}
	params.Add("q", city)
	params.Add("appid", w.apiKey)
	params.Add("units", "metric") // Use Celsius
	params.Add("lang", "tr")      // Turkish language

	reqURL := fmt.Sprintf("%s?%s", baseURL, params.Encode())
	w.logger.Debug("Weather API URL", "url", reqURL)

	// Create HTTP request
	req, err := http.NewRequestWithContext(ctx, "GET", reqURL, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	w.logger.Debug("Fetching weather data", "city", city, "url", reqURL)

	// Execute request
	resp, err := w.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	// Read response body
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	// Check HTTP status code
	if resp.StatusCode != http.StatusOK {
		var errorResp struct {
			Cod     string `json:"cod"`
			Message string `json:"message"`
		}
		if err := json.Unmarshal(body, &errorResp); err != nil {
			return nil, fmt.Errorf("API request failed with status %d: %s", resp.StatusCode, string(body))
		}
		return nil, fmt.Errorf("API request failed: %s", errorResp.Message)
	}

	// Parse response
	var weatherData models.WeatherData
	if err := json.Unmarshal(body, &weatherData); err != nil {
		return nil, fmt.Errorf("failed to parse weather data: %w", err)
	}

	w.logger.Info("Successfully fetched weather data",
		"city", weatherData.Name,
		"temperature", weatherData.Main.Temp,
		"description", weatherData.Weather[0].Description)

	return &weatherData, nil
}

File: MCPSERVERNET/MCPServerNet.Core/Services/WeatherService.cs
using System.Net.Http.Json;
using System.Text.Json;
using Microsoft.Extensions.Logging;
using MCPServerNet.Core.Models;

namespace MCPServerNet.Core.Services;

public class WeatherService : IWeatherService
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<WeatherService> _logger;
    private readonly string _apiKey;

    public WeatherService(HttpClient httpClient, ILogger<WeatherService> logger, string apiKey)
    {
        _httpClient = httpClient;
        _logger = logger;
        _apiKey = apiKey;
    }

    public async Task<WeatherData> GetWeatherDataAsync(string city, CancellationToken cancellationToken = default)
    {
        var baseUrl = "https://api.openweathermap.org/data/2.5/weather";
        var queryParams = new Dictionary<string, string>
        {
            ["q"] = city,
            ["appid"] = _apiKey,
            ["units"] = "metric", // Celsius
            ["lang"] = "tr"       // Turkish language
        };

        var queryString = string.Join("&", queryParams.Select(kvp => $"{kvp.Key}={Uri.EscapeDataString(kvp.Value)}"));
        var requestUrl = $"{baseUrl}?{queryString}";

        _logger.LogDebug("Weather API URL: {Url}", requestUrl);
        _logger.LogDebug("Fetching weather data for city: {City}", city);

        try
        {
            var response = await _httpClient.GetAsync(requestUrl, cancellationToken);
            
            if (!response.IsSuccessStatusCode)
            {
                var errorContent = await response.Content.ReadAsStringAsync(cancellationToken);
                var errorResponse = JsonSerializer.Deserialize<ErrorResponse>(errorContent);
                
                throw new HttpRequestException(
                    $"API request failed: {errorResponse?.Message ?? response.StatusCode.ToString()}", 
                    null, 
                    response.StatusCode);
            }

            var weatherData = await response.Content.ReadFromJsonAsync<WeatherData>(
                cancellationToken: cancellationToken);

            if (weatherData == null)
            {
                throw new InvalidOperationException("Failed to deserialize weather data");
            }

            _logger.LogInformation(
                "Successfully fetched weather data for city: {City}, temperature: {Temperature}, description: {Description}",
                weatherData.Name,
                weatherData.Main.Temp,
                weatherData.Weather.FirstOrDefault()?.Description);

            return weatherData;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to get weather data for city: {City}", city);
            throw;
        }
    }

    private class ErrorResponse
    {
        public string Cod { get; set; } = string.Empty;
        public string Message { get; set; } = string.Empty;
    }
} 